README.md

```diff
@@ -1,7 +1,145 @@
 
+1. Set up
+2. Matryoshka Linear Layer
+3. Training ResNet50 Models
+4. Evaluating Trained Models
+5. Evaluating pre-trained checkpoints
+5. Retrieval Performance
+
+Lastly, we also provide code to generate the figures mentioned in our paper. Please have a look at thee `Visualization/` folder. 
+## Set Up
+First pip install the requirements file in this directory:
+```
+pip install -r requirements.txt
+```
+We use PyTorch Distributed Data Parallel shared over 2 A100 GPUs and FFCV for the dataloader. Please make sure to scale the learning rate according to the #GPUs used. 
+
+### Preparing the dataset.
+
+Generate ImageNet FFCV dataset from standard imagenet train set with the following command (`IMAGENET_DIR` should point to a PyTorch style [ImageNet dataset](https://github.com/MadryLab/pytorch-imagenet-dataset)):
+
+```bash
+# Required environmental variables for the script:
+export IMAGENET_DIR=/path/to/pytorch/format/imagenet/directory/
+export WRITE_DIR=/your/path/here/
+
+# Serialize images with:
+# - 500px side length maximum
+# - 50% JPEG encoded, 90% raw pixel values
+# - quality=90 JPEGs
+.train_and_eval/train/write_imagenet.sh 500 0.50 90
+```
+## Matryoshka Linear Layer
+`MRL.py` provides MRL linear layer, and it can be instantiated very easily using the following
+```
+nesting_list=[8, 16, 32, 64, 128, 256, 512, 1024, 2048]
+fc_layer = MRL_Linear_Layer(nesting_list, num_classes=1000, efficient=efficient)
+```
+Where, `nesting_list` is the list of dimensions we want to nest on, `num_classes` is the number of output feature and efficient flag if we want to do MRL-E.
+
+## Training ResNet50 models
+First of all change the directory to `train_and_eval/train`. While training, we dump  model ckpts, training logs and run command inside a folder. 
+
+### Training Fixed Feature Baseline
+
+```bash 
+export CUDA_VISIBLE_DEVICES=0,1,2,3
+
+python train_imagenet.py --config-file rn50_configs/rn50_40_epochs.yaml --model.fixed_feature=2048 --data.train_dataset=/home/jupyter/dataset-ffcv/train_500_0.50_90.ffcv --data.val_dataset=/home/jupyter/dataset-ffcv/uncompressed/val_500_uncompressed.ffcv --data.num_workers=12 --data.in_memory=1 --logging.folder=trainlogs --logging.log_level=1 --dist.world_size=2 --training.distributed=1 --lr.lr=0.425
+```
+
+### Training MRL model
+
+```bash 
+export CUDA_VISIBLE_DEVICES=0,1,2,3
+
+python train_imagenet.py --config-file rn50_configs/rn50_40_epochs.yaml --model.mrl=1 --data.train_dataset=/home/jupyter/dataset-ffcv/train_500_0.50_90.ffcv --data.val_dataset=/home/jupyter/dataset-ffcv/uncompressed/val_500_uncompressed.ffcv --data.num_workers=12 --data.in_memory=1 --logging.folder=trainlogs --logging.log_level=1 --dist.world_size=2 --training.distributed=1 --lr.lr=0.425
```

Comment:
replace `/home/jupyter/dataset-ffcv/` and `/home/jupyter/dataset-ffcv/uncompressed/` with `WRITE_DIR`